File: fantasy/vor.py
"""
fantasy.vor — pure VOR (value over replacement) + tiering, computed from a league's actual
roster requirements and size (never a fixed 12-team/PPR assumption). Everything here is a
function of already-scored players (fantasy_points from scoring.py) plus league shape -- no
I/O, matching valuation.py's boundary.
"""
from __future__ import annotations

from typing import Sequence

# Sleeper roster_positions entries that are drafted starters (excludes bench/IR/taxi).
FLEX_ELIGIBLE: dict[str, set[str]] = {
    "FLEX": {"RB", "WR", "TE"},
    "SUPER_FLEX": {"QB", "RB", "WR", "TE"},
    "REC_FLEX": {"WR", "TE"},
    "WRRB_FLEX": {"RB", "WR"},
}
BENCH_SLOTS = {"BN", "IR", "TAXI"}


def starter_counts(roster_positions: Sequence[str], league_size: int) -> dict[str, float]:
    """Expected number of drafted starters at each base position (QB/RB/WR/TE/K/DEF) across
    the whole league, splitting FLEX-type slots evenly across their eligible positions -- an
    explicit, documented approximation (not a simulation of real draft behavior), same spirit
    as valuation.py preferring a disclosed heuristic over a hidden model."""
    counts: dict[str, float] = {}
    for slot in roster_positions:
        slot = slot.upper()
        if slot in BENCH_SLOTS:
            continue
        if slot in FLEX_ELIGIBLE:
            eligible = FLEX_ELIGIBLE[slot]
            for pos in eligible:
                counts[pos] = counts.get(pos, 0.0) + 1.0 / len(eligible)
        else:
            counts[slot] = counts.get(slot, 0.0) + 1.0
    return {pos: n * league_size for pos, n in counts.items()}


def replacement_levels(scored_players: list[dict], roster_positions: Sequence[str],
                       league_size: int) -> dict[str, float]:
    """The projected fantasy_points of the last startable player at each position -- the
    'replacement level' VOR is measured against. A position with zero counted starters (e.g.
    a league with no K slot) has no replacement level and is excluded."""
    counts = starter_counts(roster_positions, league_size)
    by_pos: dict[str, list[float]] = {}
    for p in scored_players:
        by_pos.setdefault(p["position"], []).append(p["fantasy_points"])
    levels: dict[str, float] = {}
    for pos, n in counts.items():
        pts = sorted(by_pos.get(pos, []), reverse=True)
        if not pts:
            continue
        idx = max(min(int(round(n)) - 1, len(pts) - 1), 0)
        levels[pos] = pts[idx]
    return levels


def compute_vor(scored_players: list[dict], roster_positions: Sequence[str],
                league_size: int) -> list[dict]:
    """Attach `vor` (fantasy_points - replacement level for that position) to a copy of every
    scored player, sorted descending. Players at a position with no replacement level (not
    started anywhere in this league's roster shape) get vor=None and sort last."""
    levels = replacement_levels(scored_players, roster_positions, league_size)
    out = []
    for p in scored_players:
        row = dict(p)
        level = levels.get(p["position"])
        row["vor"] = round(p["fantasy_points"] - level, 2) if level is not None else None
        out.append(row)
    out.sort(key=lambda r: (r["vor"] is None, -(r["vor"] or 0)))
    return out


def assign_tiers(vor_ranked: list[dict], gap_threshold: float = 0.75) -> list[dict]:
    """Group into tiers via natural gaps in VOR: a new tier starts whenever the drop to the
    next player exceeds `gap_threshold` standard deviations above the mean consecutive gap --
    an explicit, disclosed rule (not a clustering black box), per the task's "natural gaps"
    ask. Players with vor=None all land in one trailing tier."""
    ranked = [p for p in vor_ranked if p.get("vor") is not None]
    unranked = [p for p in vor_ranked if p.get("vor") is None]

    if len(ranked) < 2:
        for p in ranked:
            p["tier"] = 1
        for p in unranked:
            p["tier"] = 2 if ranked else 1
        return ranked + unranked

    gaps = [ranked[i]["vor"] - ranked[i + 1]["vor"] for i in range(len(ranked) - 1)]
    mean_gap = sum(gaps) / len(gaps)
    variance = sum((g - mean_gap) ** 2 for g in gaps) / len(gaps)
    cutoff = mean_gap + gap_threshold * (variance ** 0.5)

    tier = 1
    ranked[0]["tier"] = tier
    for i, gap in enumerate(gaps):
        if gap > cutoff and gap > 0:
            tier += 1
        ranked[i + 1]["tier"] = tier
    for p in unranked:
        p["tier"] = tier + 1
    return ranked + unranked

File: static/api/fantasy/_vor.js
// Port of fantasy/vor.py -- pure VOR + tiering. Keep in sync with that file.

const FLEX_ELIGIBLE = {
  FLEX: new Set(["RB", "WR", "TE"]),
  SUPER_FLEX: new Set(["QB", "RB", "WR", "TE"]),
  REC_FLEX: new Set(["WR", "TE"]),
  WRRB_FLEX: new Set(["RB", "WR"]),
};
const BENCH_SLOTS = new Set(["BN", "IR", "TAXI"]);

function starterCounts(rosterPositions, leagueSize) {
  const counts = {};
  for (let slot of rosterPositions) {
    slot = slot.toUpperCase();
    if (BENCH_SLOTS.has(slot)) continue;
    const eligible = FLEX_ELIGIBLE[slot];
    if (eligible) {
      for (const pos of eligible) counts[pos] = (counts[pos] || 0) + 1 / eligible.size;
    } else {
      counts[slot] = (counts[slot] || 0) + 1;
    }
  }
  const out = {};
  for (const [pos, n] of Object.entries(counts)) out[pos] = n * leagueSize;
  return out;
}

function replacementLevels(scoredPlayers, rosterPositions, leagueSize) {
  const counts = starterCounts(rosterPositions, leagueSize);
  const byPos = {};
  for (const p of scoredPlayers) (byPos[p.position] = byPos[p.position] || []).push(p.fantasy_points);
  const levels = {};
  for (const [pos, n] of Object.entries(counts)) {
    const pts = (byPos[pos] || []).slice().sort((a, b) => b - a);
    if (!pts.length) continue;
    const idx = Math.max(Math.min(Math.round(n) - 1, pts.length - 1), 0);
    levels[pos] = pts[idx];
  }
  return levels;
}

function computeVor(scoredPlayers, rosterPositions, leagueSize) {
  const levels = replacementLevels(scoredPlayers, rosterPositions, leagueSize);
  const out = scoredPlayers.map(p => {
    const level = levels[p.position];
    const vor = level !== undefined ? Math.round((p.fantasy_points - level) * 100) / 100 : null;
    return { ...p, vor };
  });
  out.sort((a, b) => {
    const aNull = a.vor === null, bNull = b.vor === null;
    if (aNull !== bNull) return aNull ? 1 : -1;
    return (b.vor || 0) - (a.vor || 0);
  });
  return out;
}

function assignTiers(vorRanked, gapThreshold = 0.75) {
  const ranked = vorRanked.filter(p => p.vor !== null);
  const unranked = vorRanked.filter(p => p.vor === null);
  if (ranked.length < 2) {
    ranked.forEach(p => { p.tier = 1; });
    unranked.forEach(p => { p.tier = ranked.length ? 2 : 1; });
    return ranked.concat(unranked);
  }
  const gaps = [];
  for (let i = 0; i < ranked.length - 1; i++) gaps.push(ranked[i].vor - ranked[i + 1].vor);
  const meanGap = gaps.reduce((a, b) => a + b, 0) / gaps.length;
  const variance = gaps.reduce((a, g) => a + (g - meanGap) ** 2, 0) / gaps.length;
  const cutoff = meanGap + gapThreshold * Math.sqrt(variance);
  let tier = 1;
  ranked[0].tier = tier;
  for (let i = 0; i < gaps.length; i++) {
    if (gaps[i] > cutoff && gaps[i] > 0) tier += 1;
    ranked[i + 1].tier = tier;
  }
  unranked.forEach(p => { p.tier = tier + 1; });
  return ranked.concat(unranked);
}

module.exports = { FLEX_ELIGIBLE, BENCH_SLOTS, starterCounts, replacementLevels, computeVor, assignTiers };
